Ensure we catch all semver pre-release versions

jonathanrainer opened the pull request:
At present we're only capturing versions with a specific set of names encoded into the tags, which caused an incident when an unexpected pre-release SemVer tag was used. This regex should ensure we capture all pre-release tags.

Tested the regex using Regex101 against some exemplars (including that which caused the incident) and it matches and doesn't match as it needs to

jonathanrainer committed:
Ensure we catch all semver pre-release versions

At present we're only capturing versions with a specific set of names
encoded into the tags, which caused an incident when an unexpected
pre-release SemVer tag was used. This regex should ensure we capture all
pre-release tags.

jonathanrainer committed:
Fix escaping

abernix reviewed (commented):
why not literally _just_ check for a dash? If it has a dash in it, it has to be pre-release per spec, no? 

https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions

it would just be easier for me to review 😀

jonathanrainer committed:
Second attempt at regex

We don't need to use non-capturing groups because we only care about
a match, and further our formulation of digits was incorrect so have
gone back to the semver standard's regex, with some changes.

jonathanrainer committed:
Ensure release workflow only runs for semver tags

This prevents someone pushing a random tag and that triggering a
release, even if it starts with v

dylan-apollo reviewed (approved):
Looks good, thanks! Maybe we should make our own orb at some point so we can share this between projects
